clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java

```diff
@@ -241,13 +241,19 @@ public void onSuccess(ClientResponse resp) {
 
                                 for (Map.Entry<TopicPartition, FetchResponse.PartitionData<Records>> entry : response.responseData().entrySet()) {
                                     TopicPartition partition = entry.getKey();
-                                    long fetchOffset = data.sessionPartitions().get(partition).fetchOffset;
-                                    FetchResponse.PartitionData<Records> fetchData = entry.getValue();
-
-                                    log.debug("Fetch {} at offset {} for partition {} returned fetch data {}",
-                                            isolationLevel, fetchOffset, partition, fetchData);
-                                    completedFetches.add(new CompletedFetch(partition, fetchOffset, fetchData, metricAggregator,
-                                            resp.requestHeader().apiVersion()));
+                                    FetchRequest.PartitionData requestData = data.sessionPartitions().get(partition);
+                                    if (requestData == null) {
+                                        // Received fetch response for missing session partition
+                                        throw new IllegalStateException("Received response for partition " + partition + " which is not part of the session");
```

Comment:
I wonder if we could include any additional state from the session itself to get closer to the root of the problem. For example, was this a response to a full fetch request?

Reply:
Add more information. Read the new code twice as none of the tests exercise the formatting logic.